docs/decisions/0015-serving-static-assets.rst

```diff
@@ -0,0 +1,154 @@
+
+  Setting these values for each request has proved problematic because some (but not all) S3-compatible storage services (including S3 itself) only support setting those headers for each request if you issue a signed GET request, which then gets in the way of caching and introduces the probability of browsers caching expired links, leading to all kinds of annoying cache invalidation issues.
+
+  Setting the filename value at upload time also doesn't work because the same data may be referenced under different filenames by different Components or even different versions of the same Component.
+
+Application Requirements
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Relative links between assets must be preserved.**
+  Assets may reference each other in relative links, e.g. a JavaScript file that references images or other JavaScript files. That means that our solution cannot require querystring-based authorization tokens in the style of S3 signed URLs, since asset files would have no way to encode those into their relative links.
+
+**Multiple versions of the asset should be available at the same time.**
+  Our system should be able to serve at minimum the current draft and published versions of an asset. Ideally, it should be able to serve any version of an asset. This is a departure from the way Studio and the LMS currently handle files and uploads, since there is currently no versioning at all–assets exist in a flat namespace at the course level and are immediately published.
+
+Security Requirements
+~~~~~~~~~~~~~~~~~~~~~
+
+**Assets must enforce user+file read permissions at the Learning Context level.**
+  The MongoDB GridFS backed ContentStore currently supports course-level access checks that can be toggled on and off for individual assets. Uploaded assets are public by default, and can be downloaded by anyone who knows the URL, regardless of whether or not they are enrolled in the course. They can optionally be "locked", which will restrict downloads to students who are enrolled in the course.
+
+**Assets should enforce more granular permissions at the individual Component level.**
+  An important distinction between ContentStore and v2 Content Library assets is that the latter can be directly associated with a Component. As a long term goal, we should be able to make permissions check on per-Component basis. So if a student does not have permission to view a Component for whatever reason (wrong content group, exam hasn't started, etc.), then they should also not have permission to see static assets associated with that component.
+
+  The further implication of this requirement is that *permissions checking must be extensible*. The openedx-learning repo will implement the details of how to serve an asset, but it will not have the necessary models and logic to determine whether it is allowed to.
+
+**Assets must be served from an entirely different domain than the LMS and Studio instances.**
```

Comment:
Does this requirement apply to all Open edX platforms, and not just edX.org? We can't possibly expect Open edX users to register a second domain name to host their platforms. At the very least, this item should be discussed with the BTR. Personally, I would oppose such a requirement.

Reply:
@regisb: It would apply to all Open edX platforms.

For now, we'll use a different sub-domain for development purposes (so it would come across like a new service). There are still some long term security risks with that, but it's better than the current state of things. I'll put a writeup together on security tradeoffs before bringing this to the BTR and likely the security working group. We won't add a hard requirement of a new domain for assets before doing that feedback cycle.

That being said, right now I'm really trying to unblock some folks on libraries work, and then I'm taking time off to go back to my hometown for the first time since the pandemic. It's likely that I won't have the time to go through this process until next month.

Thank you for your feedback.

Reply:
FYI @bradenmacdonald, @kdmccormick: ^

Reply:
> I'll put a writeup together on security tradeoffs before bringing this to the BTR and likely the security working group. 

Thanks Dave, I'll keep an eye out for that.

> We can't possibly expect Open edX users to register a second domain name to host their platforms. 

This is a good point, and it makes me wonder if a sibling subdomain would be sufficient for security. For example, if a site operator today had:

* example.org  # lms
* studio.example.org  # cms
* apps.example.org  # mfes

could we move them to:

* assets.example.org  # assets
* lms.example.org   # lms
* studio.example.org  # cms
* apps.example.org  # mfes
* example.org/*  # caddy wildcard redirect to lms.example.org/*, just to preserve existing LMS URLs.

Reply:
A sibling subdomain only works if you are extremely careful to never set cookies onto the root domain. If you look at tutor today, for example, many cookies like `lms_sessionid` get set onto the root `.local.edly.io` domain. While it may be possible to configure Open edX to work without root domain cookies, there is not really any secure way to enforce this "don't set cookies on the root domain" as a policy, so it remains a potential security issue. Plus it works both ways - if you use a subdomain domain for untrusted content, that untrusted content can _set_ cookies onto your root domain. (Unless it's a [public suffix](https://publicsuffix.org/) like `opencraft.hosting`.)

It's much safer to use a completely unrelated domain.

> We can't possibly expect Open edX users to register a second domain name to host their platforms.

To be clear, _registering_ a second domain name is usually not going to be necessary except for aesthetic reasons.

Almost all production deployments happen on the cloud and virtually all cloud providers provide free domain names on public suffixes, e.g. `d111111abcdef8.cloudfront.net` or `x.s3.us-west-2.amazonaws.com` (AWS), `x.r2.dev` or `x.workers.dev` (CloudFlare), etc. Any of these free domains will suffice.